### DISCUSSION TITLE
vector::push_back의 두 가지 Time Complexity 메모

### DISCUSSION THREAD

**Discussion by Woon-2** [MyeongWoonJang/cpp-study#182] [General]
vector::push_back

Asymptotic Time Complexity
f: n -> n만큼의 입력 데이터에 대한 시간 소모량
f(n) <= k * g(n)이면 f -> O(g(n))

vector::push_back은
f(n) <= k * n을 만족하므로 O(n)

Amortized Time Complexity
(Asymptotic Time Complexity가 1회의 수행으로 일어날 수 있는 최악의 경우를 따지는 반면,
Amortized Time Complexity는 m회의 수행으로 일어날 수 있는 최악의 경우를 따짐 -> 평균을 내서)

T: m -> 함수를 m번 실행했을 때의 시간 소모량
T(m) = t0 + t1 + t2 + ... + tm (수열 tn은 함수를 1회 실행했을 때 시간 소모량)
T(m)/m = (t0 + t1 + t2 + ... + tm)/m = O(g(m))이면 T -> O(g(m))

vector::push_back의 경우
T(m) = (메모리 재할당 X인 경우의 시간) + (메모리 재할당 O인 경우의 시간)
m번의 시행 중 log(m)번은 메모리 재할당 O
m번의 시행 중 m - log(m)번은 메모리 재할당 X
(메모리 재할당 X인 경우의 시간) = { m - log(m) } * 1 = m - log(m)
(메모리 재할당 O인 경우의 시간) = 2^1 + 2^2 + 2^3 + ... + 2^log(m) // (참고: 2^log(m) = m)
= [ 2^{ log(m)+1 }-2 ] / (2 - 1) = 2^{ log(m)+1 }-2 = 2^log(m) * 2 - 2 = 2*m-2 // (참고: 등비수열의 합)
따라서 T(m) = m - log(m) + 2*m - 2 = 3*m - log(m) + 2
T(m)/m = 3 - log(m)/m + 2/m
3 - log(m)/m + 2/m < k * 1을 만족하므로 O(1)

**Comment by Lcat12345** [comment]
f(n) <= k * n을 만족하므로 O(n) 이여서
3 - log(m)/m + 2/m < k * 1을 만족하므로 O(1) 이 나온건가요

**Reply by Woon-2** [reply]
넴 맞아용
